libs/auth/react/src/lib/auth/AuthProvider.tsx

```diff
@@ -0,0 +1,260 @@
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useReducer,
+  useRef,
+  ReactNode,
+  useState,
+} from 'react'
+import type { User } from 'oidc-client-ts'
+import { Location } from 'react-router-dom'
+import { getAuthSettings, getUserManager } from '../userManager'
+import { ActionType, initialState, reducer } from './Auth.state'
+import { AuthSettings } from '../AuthSettings'
+import { AuthContext } from './AuthContext'
+import { AuthErrorScreen } from './AuthErrorScreen'
+import { AuthLoadingScreen } from './AuthLoadingScreen'
+import { CheckIdpSession } from './CheckIdpSession'
+import { isDefined } from '@island.is/shared/utils'
+
+interface AuthProviderProps {
+  /**
+   * If true, Authenticator automatically starts login flow and does not render children until user is fully logged in.
+   * If false, children are responsible for rendering a login button and loading indicator.
+   * Default: true
+   */
+  autoLogin?: boolean
+  /**
+   * The base path of the application.
+   */
+  basePath: string
+  children: ReactNode
+}
+
+const getReturnUrl = (returnUrl: string, { redirectPath }: AuthSettings) => {
+  if (redirectPath && returnUrl.startsWith(redirectPath)) {
+    return '/'
+  }
+
+  return returnUrl
+}
+
+export const AuthProvider = ({
+  children,
+  autoLogin = true,
+  basePath,
+}: AuthProviderProps) => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+  const [hasError, setHasError] = useState(false)
+  const userManager = getUserManager()
+  const authSettings = getAuthSettings()
+  const monitorUserSession = !authSettings.scope?.includes('offline_access')
+  const urlRef = useRef<string>()
+  urlRef.current = `${window.location.pathname}${window.location.search}`
+
+  const signIn = useCallback(async () => {
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+
+    return userManager.signinRedirect({
+      state: getReturnUrl(urlRef.current as string, authSettings),
+    })
+    // Nothing more happens here since browser will redirect to IDS.
+  }, [dispatch, userManager, authSettings, urlRef])
+
+  const signInSilent = useCallback(async () => {
+    let user = null
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+    try {
+      user = await userManager.signinSilent()
+      dispatch({ type: ActionType.SIGNIN_SUCCESS, payload: user })
+    } catch (error) {
+      console.error('AuthProvider: Silent signin failed', error)
+      dispatch({ type: ActionType.SIGNIN_FAILURE })
+    }
+
+    return user
+  }, [userManager, dispatch])
+
+  const switchUser = useCallback(
+    async function switchUser(nationalId?: string, newLocation?: Location) {
+      if (newLocation) {
+        urlRef.current = `${newLocation.pathname}${newLocation.search}`
+      }
```

Comment:
This `newLocation` argument was added to deal with an issue with out of date React Router useLocation data.  We might want to clean it up and just use window.location directly.

We might have an issue with basePath here as well, since the Location object comes from react router, probably without a basepath.

Reply:
Good spot